### PR TITLE
DR-1123 Ensure stewards, custodians, and ingesters can read dataset-level files

### PR DESCRIPTION
**Pull request by danxmoran**
This is important for a Monster use-case (using the TDR to hold some intermediate data, with the option of "reprocessing" data from midway through the pipeline by reading from the TDR files).

I might have completely broken the integration test auth logic, opening a draft to see what happens!

### CODE REVIEW

File: src/test/java/bio/terra/common/auth/AuthService.java
Comment (danxmoran) [inline src/test/java/bio/terra/common/auth/AuthService.java:61]: I could probably revert all the changes in this file and do the `GoogleCredentials` setup manually in the test where I need it. I'm not sure if it's the cause of the failures in the connected tests, though

File: src/main/java/bio/terra/service/filedata/flight/ingest/IngestFileAuthzStep.java
Comment (danxmoran) [inline src/main/java/bio/terra/service/filedata/flight/ingest/IngestFileAuthzStep.java:20]: Is there any chance this could be causing serialization issues for the step?

File: src/main/java/bio/terra/service/filedata/flight/ingest/FileIngestFlight.java
Comment (ddietterich) [inline src/main/java/bio/terra/service/filedata/flight/ingest/FileIngestFlight.java:49]: this is the code for the single file load. We are trying to make this obsolete.
Do you really need to use this method any more?

File: src/main/java/bio/terra/service/iam/sam/SamIam.java
Comment (ddietterich) [inline src/main/java/bio/terra/service/iam/sam/SamIam.java:198]: I don't think you want to sync the policies at this point. The policies do not need to be updated.
All you need are the group names. Maybe we need to save those from when we create the dataset.

File: src/test/java/bio/terra/common/auth/AuthService.java
Comment (ddietterich) [inline src/test/java/bio/terra/common/auth/AuthService.java:55]: This looks like it only used in one place. Do we really need another method?

**Review by ddietterich — CHANGES_REQUESTED** [review]
This doesn't look right to me, and it doesn't cover all of the load code.

File: src/main/java/bio/terra/service/filedata/flight/ingest/FileIngestFlight.java
Comment (danxmoran) [reply src/main/java/bio/terra/service/filedata/flight/ingest/FileIngestFlight.java:49]: We haven't shifted off of it for the one project blocked by this issue, but I was also assuming the bulk load ended up delegating to this flight at some point.

File: src/main/java/bio/terra/service/iam/sam/SamIam.java
Comment (danxmoran) [reply src/main/java/bio/terra/service/iam/sam/SamIam.java:198]: I had considered that as well, I wasn't sure if there'd be a chance that Sam switched up the groups (I guess that's unlikely since they don't have a mechanism to update the groups at every use-site).

The naming of the method here is a bit misleading though; `syncStatus` is fetching the latest status of synchronization (GET), not triggering a sync of statuses (POST).

File: src/test/java/bio/terra/common/auth/AuthService.java
Comment (danxmoran) [reply src/test/java/bio/terra/common/auth/AuthService.java:55]: Probably not, I'm just going to revert this file and make a more localized change.

File: src/main/java/bio/terra/service/iam/sam/SamIam.java
Comment (danxmoran) [reply src/main/java/bio/terra/service/iam/sam/SamIam.java:198]: My main concern for the track-policies-in-the-DB case is figuring out how we migrate existing datasets.

File: src/main/java/bio/terra/service/filedata/flight/ingest/FileIngestFlight.java
Comment (danxmoran) [reply src/main/java/bio/terra/service/filedata/flight/ingest/FileIngestFlight.java:49]: I see FileIngestWorkerFlight now, I'll update there too.

Comment (ddietterich) [reply src/main/java/bio/terra/service/filedata/flight/ingest/FileIngestFlight.java:49]: No, processing is incompatible between the two flights. We really need to get rid of this one. We have not been maintaining it in conjunction with the other ingest flights.

File: src/main/java/bio/terra/service/filedata/flight/ingest/IngestFileAuthzStep.java
Comment (ddietterich) [inline src/main/java/bio/terra/service/filedata/flight/ingest/IngestFileAuthzStep.java:46]: There is a bug at this statement. The file object is not complete and those so this fails with found not found.
Further, it might be possible to provide read access at the bucket level, instead of setting it on each file.

Comment (danxmoran) [reply src/main/java/bio/terra/service/filedata/flight/ingest/IngestFileAuthzStep.java:46]: Conceptually I like the idea of bucket-level access for this use-case, but wouldn't that require shifting from a bucket per profile to a bucket per profile/dataset?

Thanks for pointing out the bug. Assuming we can keep changes targeted for now, would a valid fix be to move this step to after the `IngestFileFileStep`?

File: src/main/java/bio/terra/service/filedata/flight/ingest/FileIngestFlight.java
Comment (danxmoran) [reply src/main/java/bio/terra/service/filedata/flight/ingest/FileIngestFlight.java:49]: I'm +1 in general to ripping out the single-file case and migrating our one use of it, but would rather not block the IAM change on it.

File: src/main/java/bio/terra/service/iam/sam/SamIam.java
Comment (danxmoran) [reply src/main/java/bio/terra/service/iam/sam/SamIam.java:198]: Hmmm but even assuming we still ask Sam for the group emails each time, this should probably be broken into multiple steps for the bulk-load case:
1. Step that gets the policy emails from somewhere and sticks them in the flight map
2. Step that sets the ACLs for the policy emails

File: src/main/java/bio/terra/service/filedata/flight/ingest/IngestFileAuthzStep.java
Comment (ddietterich) [reply src/main/java/bio/terra/service/filedata/flight/ingest/IngestFileAuthzStep.java:20]: I don't think so.

**Comment by ddietterich** [comment]
I think the right solution to this problem is to the permissions on the bucket. I suggest we do that instead of this approach. For now we can just apply it to the one bucket and make a separate ticket for making a bucket for each dataset.

In the context of dataset creation, we already have all of the names of the policies so we would not have to do this retrieval.
@thathert what do you think?

**Comment by danxmoran** [comment]
Bucket-level permissions would definitely unblock us in the short-term. We only really need steward-level access, if you'd rather not set profile-wide ACLs for any custodians or ingesters.

**Comment by ddietterich** [comment]
This would just be a short term state until we can use bucket per dataset

**Comment by danxmoran** [comment]
Slacked with @thathert and got a virtual thumbs up to (manually):
1. Add the dev stewards policy as an object viewer on broad-jade-dev-data-bucket
2. Add the "prod" stewards policy as an object viewer on broad-datarepo-terra-prod-cgen-bucket

That will cover our immediate use-cases for ClinVar. Closing this PR.

**Pull request closed without merging**
